[xutility] Modernize _Equal_memcmp_is_safe to use variable templates

miscco opened the pull request:
This moves the helper structs `_Equal_memcmp_is_safe`  and `_Equal_memcmp_is_safe_helper` to use variable templates rather than structs.

This is a precursor to #819

miscco commented:
*Just one more round* is great for civilization bot not good for PRs. This should now be ready for initial review.

mnatsuhara commented on stl/inc/xutility:
```diff
@@ -4767,7 +4757,7 @@ _NODISCARD bool equal(const _InIt1 _First1, const _InIt1 _Last1, const _InIt2 _F
     return _Equal_unchecked(_UFirst1, _ULast1, _UFirst2, _Pass_fn(_Pred));
 }
-#endif // _HAS_IF_CONSTEXPR
+#endif // !_HAS_IF_CONSTEXPR
```
Is this what we typically do?  If you look at the example just below (line 4793-4871 before changes), the `#endif` is commented with the original condition (not negated).  Doing a quick search throughout the code base also shows that the original comment follows our convention.

miscco replied:
As far as I know there was a discussion early after open sourcing and I have the memory that the majority was in favor of 

```cpp
#if foo
#else // ^^^ foo / !foo
#endif // !foo
```

The reason was that for longer clauses such as this just reading `#endif // foo` would lead you to believe that foo is active or require you to check the start of the block.

That said sleep was rather short this year so I doubt any memory I have

StephanTLavavej replied:
This is #351 - existing code has accumulated inconsistent patterns. I'm not sure how we ended up with the "hybrid arrow" examples that @mnatsuhara pointed out - I think they evolved from our classic pattern of

```cpp
#if _HAS_CATS
#else // _HAS_CATS
#endif // _HAS_CATS

#ifdef __cpp_lib_kittens
#else // __cpp_lib_kittens
#endif // __cpp_lib_kittens
```

by replacing just the middle portion with arrows.

My personal opinion: Aside from the question of when comments can be omitted, I'd prefer for us to standardize on consistent arrow comments:

```cpp
#if _HAS_CATS
#else // ^^^ _HAS_CATS ^^^ / vvv !_HAS_CATS vvv
#endif // ^^^ !_HAS_CATS ^^^

#ifdef __cpp_lib_kittens
#else // ^^^ defined(__cpp_lib_kittens) ^^^ / vvv !defined(__cpp_lib_kittens) vvv
#endif // ^^^ !defined(__cpp_lib_kittens) ^^^
```

(I'm not very concerned with whether ` ^^^ / vvv ` or just ` / ` appears in the middle; I like the former slightly more.) I prefer this because it's the clearest in dense regions of preprocessor logic, and it's also unambiguous about whether the final comment should say `_HAS_CATS` or `!_HAS_CATS` (because it points to a specific section of code).

I could see an argument that "of course an `#endif` closes a section of code, so we just need to consistently identify what we closed", which would argue for `!_HAS_CATS` here.

Somewhat related: I believe we've accumulated some inconsistency about how to refer to `#ifdef` (and `#if defined(MEOW)` etc.) conditions in comments - I have seen some comments switch from the `defined(MEOW)` / `!defined(MEOW)` domain to the `MEOW` / `!MEOW` domain which I find confusing. This is because of the danger of actually confusing the definedness versus valueness domains in the test (and the compiler warning about `#if NOT_DEFINED` being treated as `#if 0`).

There are "only" 444 occurrences of `#else` in `stl/inc`, but 2049 `#endif`s - tedious to audit, but still easier than bracing all control flow, which we finally did and it was awesome. I should probably just submit a PR to prevent this from burning up valuable time in the future. :fire:

miscco replied:
I have to agree that we should be consistent. As I had to rebase anyway I removed that change so that it can be tackled with all the other comments in a consistent fashion

miscco replied:
Maybe this would be a "good second issue"

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4650,55 +4650,51 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
-}
+// return equal optimization category for pointers
+template <class _Elem1, class _Elem2, class _Pr>
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe<_Elem1*, _Elem2*, _Pr> =
```
This is introducing a potential usability issue. Previously, `_Equal_memcmp_is_safe(ptr1, ptr2, pred)` ignored top-level `const` on the pointers, due to `_Obj1* const&, _Obj2* const&`. Now, `_Equal_memcmp_is_safe<_Elem1*, _Elem2*, _Pr>` won't activate for `int* const`. This could be formed if an algorithm applies top-level `const` to an unwrapped iterator and then uses `decltype` which will pick up that `const`.

Should we accept top-level `const` here, and add tests to `VSO_0180469_ptr_cat`? Long ago, I would have introduced another layer of metaprogramming to handle this - but it should be sufficient to use your `enable_if_t` trick here, enabling this specialization for `is_pointer_v` (the primary type categories ignore top-level const) and then using `remove_pointer_t` before `remove_const_t` below.

miscco replied:
I will have to think about this a bit more. I thought that the `remove_const_t` was sufficient but it seems it is not.

In any case I should add some tests.

StephanTLavavej replied:
The `remove_const_t` there is ignoring constness on the elements, but not top-level const on the pointers.

miscco replied:
So I wrote some small tests

```cpp
    // Top level const should work
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1* const, Elem2*, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1*, Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1* const, Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1* const, const Elem2*, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1*, const Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, Elem1* const, const Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, const Elem1* const, Elem2*, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, const Elem1*, Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, const Elem1* const, Elem2* const, equal_to<>>();
    test_case_Equal_memcmp_is_safe_helper<Expected, const Elem1* const, const Elem2* const, equal_to<>>();
```

However, I did not get it to pass yet. Is there some secret way to ask the compiler which templates are instantiated?

StephanTLavavej replied:
The technique I use is getting a template to instantiate some other template that I've declared but not defined. For example:

```cpp
template <typename A, typename B, typename C> struct Meow;

template <typename T, typename U> auto add(T t, U u) -> decltype(t + u) {
    Meow<T, U, decltype(t + u)>::print();
    return t + u;
}

int main() {
    add(static_cast<unsigned char>(1), static_cast<unsigned short>(2));
}
```

MSVC emits:
```
error C2027: use of undefined type 'Meow<T,U,int>'
        with
        [
            T=unsigned char,
            U=unsigned short
        ]
```

Clang emits:
```
error: implicit instantiation of undefined template 'Meow<unsigned char, unsigned short, int>'
```

miscco replied:
Thanks to your help this is now working

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4708,7 +4704,7 @@ _NODISCARD _CONSTEXPR20 bool equal(const _InIt1 _First1, const _InIt1 _Last1, co
     const auto _ULast1 = _Get_unwrapped(_Last1);
     auto _UFirst2      = _Get_unwrapped_n(_First2, _Idl_distance<_InIt1>(_UFirst1, _ULast1));
-    if constexpr (decltype(_Equal_memcmp_is_safe(_UFirst1, _UFirst2, _Pred))::value) {
+    if constexpr (_Equal_memcmp_is_safe<_InIt1, _InIt2, _Pr>) {
```
This transformation is incorrect, preventing the optimization from activating in certain situations. "Unwrapping" iterators (e.g. from `vector<T>::iterator` to `T*`) serves two purposes:

* it increases performance by front-loading iterator debugging checks followed by direct use of raw pointers, and
* it makes metaprogramming optimizations easier.

Here, `_Equal_memcmp_is_safe` recognizes raw pointers, and then inspects the elements to see whether they're eligible for `memcmp`. It doesn't know how to recognize `vector::iterator`, but that's fine, because `vector::iterator` knows how to unwrap to a raw pointer.

Here, we were previously passing the unwrapped `_UFirst1, _UFirst2`, inspecting the unwrapped iterators. Now, you're passing the original `_InIt1, _InIt2`, which are still wrapped. This will activate the optimization only if the user actually passed raw pointers, not if they passed `vector::iterator` etc.

miscco replied:
I am absolutely amazed that you found that bug. 

I changed it to `_Equal_memcmp_is_safe<decltype(_UFirst), decltype(_UFirst2), _Pr>`

that should pass the unwrapped raw pointer to `_Equal_memcmp_is_safe`

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4729,8 +4725,8 @@ _NODISCARD _CONSTEXPR20 bool equal(const _InIt1 _First1, const _InIt1 _Last1, co
 #else // ^^^ _HAS_IF_CONSTEXPR / !_HAS_IF_CONSTEXPR vvv
-template <class _InIt1, class _InIt2, class _Pr>
-bool _Equal_unchecked1(_InIt1 _First1, const _InIt1 _Last1, _InIt2 _First2, _Pr _Pred, false_type) {
+template <class _InIt1, class _InIt2, class _Pr, enable_if_t<!_Equal_memcmp_is_safe<_InIt1, _InIt2, _Pr>, int> = 0>
+bool _Equal_unchecked(_InIt1 _First1, const _InIt1 _Last1, _InIt2 _First2, _Pr _Pred) {
```
Note: the `!_HAS_IF_CONSTEXPR` region is correctly transformed, because these are the `_Equal_unchecked` helper functions. We've already performed unwrapping in `equal()`. (This structure was necessary in the era before `decltype` when we weren't able to ask "what is the type of this iterator after unwrapping" except by introducing a helper function to perform template argument deduction, which then needed another layer of helper functions to perform tag dispatch.)

No change requested here.

StephanTLavavej commented on tests/std/tests/VSO_0180469_ptr_cat/test.cpp:
```diff
@@ -214,7 +214,7 @@ void ptr_cat_test_cases() {
+    STATIC_ASSERT(_Equal_memcmp_is_safe<Ptr1, Ptr2, Pr> == Expected);
 }
 
 template <bool Expected, class Elem1, class Elem2>
```
As mentioned in the other comment, if we make the variable template ignore top-level `const` on pointers, we should add test cases to `test_case_Equal_memcmp_is_safe()` below.

miscco replied:
Added some more tests

miscco commented on stl/inc/algorithm:
```diff
@@ -1384,8 +1384,8 @@ _NODISCARD _CONSTEXPR20 bool _Equal_rev_pred_unchecked(_InIt1 _First1, _InIt2 _F
 #else // ^^^ _HAS_IF_CONSTEXPR ^^^ // vvv !_HAS_IF_CONSTEXPR vvv
-template <class _InIt1, class _InIt2, class _Pr>
-bool _Equal_rev_pred_unchecked1(_InIt1 _First1, _InIt2 _First2, const _InIt2 _Last2, _Pr _Pred, false_type) {
+template <class _InIt1, class _InIt2, class _Pr, enable_if_t<!_Equal_memcmp_is_safe<_InIt1, _InIt2, _Pr>, int> = 0>
+bool _Equal_rev_pred_unchecked(_InIt1 _First1, _InIt2 _First2, const _InIt2 _Last2, _Pr _Pred) {
```
I was indeed unsure whether to just use tag dispatch given that it is essentially code on life support.

In the end I could not help myself ;)

miscco commented on stl/inc/xutility:
```diff
@@ -4650,55 +4650,51 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper = false;
 
+// allow memcmping same-size integral non-bool non-volatile bitwise types using equal_to<>
+// clang-format off
```
Done

miscco commented on stl/inc/xutility:
```diff
@@ -4650,55 +4650,51 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
-    : bool_constant<conjunction_v<bool_constant<sizeof(_Elem1) == sizeof(_Elem2)
-                                                && _Is_nonbool_integral<_Elem1> && _Is_nonbool_integral<_Elem2>>,
-          negation<is_volatile<_Elem1>>, negation<is_volatile<_Elem2>>,
-          // order matters here: being integral is a precondition of _Value_equality_is_bitwise_equality
```
I added the last part of the comment

miscco commented on stl/inc/xutility:
```diff
@@ -4650,55 +4650,51 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
+// allow memcmping std::byte
 template <>
-struct _Equal_memcmp_is_safe_helper<byte, byte, equal_to<>> : true_type {}; // allow memcmping std::byte
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<byte, byte, equal_to<>> = true;
```
inline comment it is. maybe not a short one though

CaseyCarter replied:
I'd prefer to use `inline` instead of `_INLINE_VAR` here.

miscco replied:
It already is ;)

miscco commented:
I am still kind of stuck here. My problem with the top level const is that I cannot really constrain the `_Equal_memcmp_is_safe` because if I use `enable_if_t` as an additional template argument, it complains about the invalid specialization. If I use `enable_if_t` on one of the arguments of the specialization it errors out because one of the template argument would be unused in the SNFINAE case.

Given that this is essentially soon to be reworked code I am actually considering just hand rolling the 4 combinations of `_Elem*` and `_Elem* const` and be done with.

StephanTLavavej commented:
Ah, I see - it's a partial specialization of a variable template, so you can't really interfere with the partial-specialization-ness.

What about the following? (I can prepare a fully worked example later if you want.) Give the primary template a default template argument of the form `bool = conjunction_v<is_pointer<ITER1>, is_pointer<ITER2>>` and make that the "true" version, which then uses `remove_pointer_t` to remove any top-level const, followed by the pointer, and then deals with the (possibly const) element type. Then, provide a partial specialization for `<ITER1, ITER2, false>` that just answers `false`. (This reverses the current pattern, because of the stylistic preference of having the primary template be used for a `true` bool.)

miscco commented:
@StephanTLavavej thanks a lot that is exactly what I needed 😲

mnatsuhara commented on stl/inc/xutility:
```diff
@@ -4650,55 +4650,53 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
-    return {};
-}
+// return equal optimization category for arbitrary iterators
+template <class _Elem1, class _Elem2, class _Pr>
```
In reference to [Stephan's previous comment](https://github.com/microsoft/STL/pull/831#discussion_r426100093), should we rename these template parameters to `_Iter1` and `_Iter2` instead of `_Elem1` and `_Elem2`?

miscco replied:
Im such a derp... Will change soon

miscco replied:
Done

miscco commented:
Sorry for the noise, I moved one of the specializations to the bottom as that will significantly reduce the churn for the followup PR with concepts and ranges::equal_to

CaseyCarter commented on stl/inc/xutility:
```diff
@@ -4650,55 +4650,53 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
+                    && !is_volatile_v<_Elem1> && !is_volatile_v<_Elem2>,
+        equal_to<>>> = _Value_equality_is_bitwise_equality<_Elem1, _Elem2>;
+
+// return equal optimization category for pointers
```
Pre-existing: I don't know what an "equal optimization category" is.
```suggestion
// Determine if it's safe to compare pointees
```

miscco replied:
Done (fixed the typo)

CaseyCarter commented on stl/inc/xutility:
```diff
@@ -4650,55 +4650,53 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
+        equal_to<>>> = _Value_equality_is_bitwise_equality<_Elem1, _Elem2>;
+
+// return equal optimization category for pointers
+template <class _Elem1, class _Elem2, class _Pr, bool = conjunction_v<is_pointer<_Elem1>, is_pointer<_Elem2>>>
```
Pre-existing: `_Elem1` and `_Elem2` are terrible names for template parameters that are iterators/pointers and not elements. We should use `_Iter1` and `_Iter2` here as well.

miscco replied:
Done

CaseyCarter commented on stl/inc/xutility:
```diff
@@ -4650,55 +4650,53 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
+    _Equal_memcmp_is_safe_helper<remove_const_t<remove_pointer_t<_Elem1>>, remove_const_t<remove_pointer_t<_Elem2>>,
+        _Pr>;
+
+// return equal optimization category for arbitrary iterators
```
Ditto "What's an 'equal optimization category'?" although here I suggest simply removing the comment.

miscco replied:
I added `// Arbitrary iterators cannot use memcmp`

CaseyCarter commented on tests/std/tests/VSO_0180469_ptr_cat/test.cpp:
```diff
@@ -239,6 +239,18 @@ void test_case_Equal_memcmp_is_safe() {
     test_case_Equal_memcmp_is_safe_helper<false, const Elem1*, const volatile Elem2*, equal_to<>>();
     test_case_Equal_memcmp_is_safe_helper<false, const volatile Elem1*, const volatile Elem2*, equal_to<>>();
 
+    // Top level const should work
```
What does "work" mean here?

miscco replied:
I moved it above and changed it to `// Top level const should not change the answer`

miscco commented on stl/inc/xstring:
```diff
@@ -487,9 +487,10 @@ struct _Char_traits_lt {
     }
 };
 
+// builtin char_traits::eq behaves like equal_to<>
```
Fixed the comment and preserved the bug

miscco commented on tests/std/tests/VSO_0180469_ptr_cat/test.cpp:
```diff
@@ -214,7 +214,7 @@ void ptr_cat_test_cases() {
 template <bool Expected, class Ptr1, class Ptr2, class Pr>
 void test_case_Equal_memcmp_is_safe_helper() {
-    STATIC_ASSERT(decltype(_Equal_memcmp_is_safe(declval<Ptr1>(), declval<Ptr2>(), declval<Pr>()))::value == Expected);
+    STATIC_ASSERT(_Equal_memcmp_is_safe<Ptr1, Ptr2, Pr> == Expected);
```
The merge conflicts....

Done

miscco commented:
So I addressed @CaseyCarter 's comments. 

Given that I rewrote a lot, I also pushed the necessary changes to enable `contiguous_iterators `and `ranges::equal_to`.

That is what you get for totally reasonable review comments. Be warned.

miscco commented:
@CaseyCarter looking at the test failures, I am wondering whether we should unwrap the iterators in the `_Equal_rev_pred_unchecked` implementation of `_Find_end_fn`

CaseyCarter commented:
> @CaseyCarter looking at the test failures, I am wondering whether we should unwrap the iterators in the `_Equal_rev_pred_unchecked` implementation of `_Find_end_fn`

They are already unwrapped; `_unchecked` in the names of algorithm helpers typically indicates that the helper expects the iterators to have been already unwrapped. The issue here is that C++20 admits contiguous iterators that are neither pointers nor library-provided that unwrap to pointers - we must use `std::to_address` to convert them to pointers before `reinterpret_cast`ing (change pushed).

CaseyCarter commented on stl/inc/xutility:
```diff
@@ -4679,6 +4679,42 @@ template <class _Elem>
+
+template <class _Iter>
+concept _Bitwise_comparable_iterator =
+    (is_pointer_v<_Iter> && !is_volatile_v<remove_pointer_t<_Iter>>) || contiguous_iterator<_Iter>;
```
`is_pointer_v<_Iter>` implies `contiguous_iterator<_Iter>`, so we need not check both, and we probably want the "but not iterators with `volatile`-qualified element type" check to apply to general iterators as well.
```suggestion
    contiguous_iterator<_Iter> && !is_volatile_v<remove_reference_t<iter_reference_t<_Iter>>>;
```

CaseyCarter commented on stl/inc/xutility:
```diff
@@ -4679,6 +4679,42 @@ template <class _Elem>
+// Determine if it's safe to compare suitable iterators
+template <_Bitwise_comparable_iterator _Iter1, _Bitwise_comparable_iterator _Iter2, class _Pr>
+inline constexpr bool _Equal_memcmp_is_safe<_Iter1, _Iter2, _Pr> =
+    _Equal_memcmp_is_safe_helper<remove_const_t<iter_value_t<_Iter1>>, remove_const_t<iter_value_t<_Iter2>>, _Pr>;
```
Since concepts short-circuit, we can simplify this to:
```c++
// Arbitrary iterators cannot use memcmp
template <class _Iter1, class _Iter2, class _Pr>
concept _Equal_memcmp_is_safe = _Bitwise_comparable_iterator<_Iter1> && _Bitwise_comparable_iterator<_Iter2>
    && _Equal_memcmp_is_safe_helper<remove_const_t<iter_value_t<_Iter1>>, remove_const_t<iter_value_t<_Iter2>>;
```
and I observe that the `is_volatile` check in `_Equal_memcmp_is_safe_helper` would be redundant with the check in `_Bitwise_comparable_iterator` if we were passing the reference types of the iterators instead of the value types which are typically cv-unqualified. Why not use:
```suggestion
// Determine if we can compare pointees with memcmp
template <class _Iter1, class _Iter2, class _Pr>
concept _Equal_memcmp_is_safe = contiguous_iterator<_Iter1> && contiguous_iterator<_Iter2>
    && _Equal_memcmp_is_safe_helper<
        remove_const_t<remove_reference_t<iter_reference_t<_Iter1>>>, 
        remove_const_t<remove_reference_t<iter_reference_t<_Iter2>>>, _Pr>;
```
and then get rid of `_Bitwise_comparable_iterator`?

miscco replied:
I am so happy right now. This is essentially my first try modulo that I screwed up with `Iter_value_t`. The whole `_Bitwise_comparable_iterator` was only to amend that.

Fun fact I already wrote a whiny comment "volatile killed my beautifull code" but thought better of it, do not whine on the internet.

So the message is loud and clear. Always whine about on the internet.

CaseyCarter commented on stl/inc/xutility:
```diff
@@ -4688,15 +4724,16 @@ _INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem1, _Elem2,
 
-// return equal optimization category for pointers
-template <class _Elem1, class _Elem2, class _Pr, bool = conjunction_v<is_pointer<_Elem1>, is_pointer<_Elem2>>>
+// Determine if it's safe to compare pointers
```
We aren't going to compare _pointers_ we're going to compare _pointees_ - the things the pointers point at.

CaseyCarter commented on tests/std/tests/VSO_0180469_ptr_cat/test.cpp:
```diff
@@ -212,61 +218,82 @@ void ptr_cat_test_cases() {
+    STATIC_ASSERT(_Equal_memcmp_is_safe<const Elem1*, Elem2* const, Pr> == Expected);
+    STATIC_ASSERT(_Equal_memcmp_is_safe<const Elem1* const, Elem2* const, Pr> == Expected);
+    STATIC_ASSERT(_Equal_memcmp_is_safe<const Elem1* const, const Elem2* const, Pr> == Expected);
+    // Iterators that are essentially pointers should not change the answer
```
Does "Iterators that are essentially pointers" mean "contiguous iterators", or something else? (Also on 250)

miscco replied:
:+1:

CaseyCarter commented on tests/std/tests/VSO_0180469_ptr_cat/test.cpp:
```diff
@@ -212,61 +218,82 @@ void ptr_cat_test_cases() {
+    STATIC_ASSERT(_Equal_memcmp_is_safe<const Elem1*, const volatile Elem2*, Pr> == false);
+    STATIC_ASSERT(_Equal_memcmp_is_safe<const volatile Elem1*, const volatile Elem2*, Pr> == false);
 
     // Iterators that are not pointers should explode
```
Non-contiguous iterators?

miscco replied:
:+1:

miscco commented on stl/inc/xutility:
```diff
@@ -4679,6 +4679,42 @@ template <class _Elem>
+#endif // __cpp_lib_byte
+
+template <class _Pr>
+concept _Equal_to_comparator = is_same_v<_Pr, equal_to<>> || is_same_v<_Pr, _RANGES equal_to>;
```
Thanks!

miscco commented:
@CaseyCarter this should now close #819

CaseyCarter commented:
> @CaseyCarter this should now close #819

Thanks for the reminder; I've attached #819 and tagged it as "work in progress."

miscco commented:
@StephanTLavavej  Thanks for improving the comments

StephanTLavavej commented:
I am working on overhauling the metaprogramming - I believe I verified the correctness of your additions yesterday, but the original design was messy to extend and it was exceeding my ability to easily reason about. I think I have something simpler that preserves your additions and enhancements, and avoids replicating too much code for concepts and non-concepts. Will add comments, test, and push soon.

miscco commented:
Given that you had trouble we can estimate hoe long I struggled to get something working.

Thanks a lot for improving it further

miscco commented on stl/inc/xutility:
```diff
@@ -4650,83 +4650,90 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
+// Here, _Elem1 and _Elem2 aren't top-level const, because we remove_const_t before using _Can_memcmp_elements.
+
+// Integral types are eligible for memcmp in very specific cases.
+// * They must be the same size. (`int == long` is eligible; `int == long long` isn't.)
```
I wonder whether you could add std::byte directly here. That would mean expanding on _Is_nonboool_integral. 

This would allow comparing std::byte with integers of compatible size. If this is the right idea I do not know

StephanTLavavej replied:
That would be an excellent idea if `std::byte` were equality comparable with integers. However, it was designed to prohibit such comparisons. It's an `enum class` which lacks the implicit conversions of a plain `enum`:
https://github.com/microsoft/STL/blob/212de15c590010c961c0e7bb4025f6fec89866c8/stl/inc/cstddef#L29
It's convertible to integers via a named function `to_integer`:
https://github.com/microsoft/STL/blob/212de15c590010c961c0e7bb4025f6fec89866c8/stl/inc/cstddef#L87-L90
But `byte == integer` is a compiler error, so we must not bypass that by calling `memcmp`.

miscco replied:
Derp, you are indeed correct. It is kind of sad that the user cannot really project to integer ranges though as we would reject the comparator

StephanTLavavej replied:
Yeah - at that point, we basically need to ask the backend to perform this optimization. (My historical understanding is that this metaprogramming was very important when the backend couldn't recognize raw loops and optimize them to `memcmp`; nowadays the metaprogramming mostly benefits non-optimized debug mode.)

miscco commented on stl/inc/xutility:
```diff
@@ -4650,83 +4650,90 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
-        _Pr>;
-
-// Arbitrary iterators cannot use memcmp
+// _Equal_memcmp_is_safe<_Iter1, _Iter2, _Pr> reports whether we can activate the memcmp optimization
```
I am wondering whether `_Equal_memcmp_is_safe` should be shortened to `_Memcmp_is_safe`. The equal doesnt really add anything

StephanTLavavej replied:
Although the name is awkwardly long, mentioning `equal` does serve a purpose. `memcmp` is a kind of spaceship comparison, reporting less/equal/greater, but this type trait specifically determines whether it can be used to replace equality. `unsigned int == unsigned int` can be performed by comparing the bytes. However, `unsigned int < unsigned int` can't be performed by comparing the bytes, because `unsigned int` is little-endian on our platforms.

`_Lex_compare_memcmp_classify` determines whether `memcmp` can be used for `lexicographical_compare`.

StephanTLavavej commented on tests/std/tests/VSO_0180469_ptr_cat/test.cpp:
```diff
@@ -212,49 +218,90 @@ void ptr_cat_test_cases() {
+    STATIC_ASSERT(_Equal_memcmp_is_safe<Elem1*, Elem2*, equal_to<const volatile Elem1>> == false);
 }
 
+template <bool Expected, class Elem1, class Elem2>
 void equal_safe_test_cases() {
```
Surprisingly, this templating is completely damaging the test. This is a compile-only test, and it's relying on non-template functions like `equal_safe_test_cases()` to instantiate the templates like `test_case_Equal_memcmp_is_safe<Expected, Elem1, Elem2>()`. This happens even though nothing attempts to call `equal_safe_test_cases()`. Adding this template prevents all of the instantiation, disabling all of the test coverage (!).

miscco replied:
So all the tests are green? Ship it!

StephanTLavavej replied:
Enabling your enhanced test coverage revealed that I misunderstood how `contiguous_iterator` works with top-level const iterators (it doesn't). I fixed the product code accordingly.

miscco commented on tests/std/tests/VSO_0180469_ptr_cat/test.cpp:
```diff
@@ -212,49 +218,90 @@ void ptr_cat_test_cases() {
+    STATIC_ASSERT(_Equal_memcmp_is_safe<const Elem1*, const volatile Elem2*, Pr> == false);
+    STATIC_ASSERT(_Equal_memcmp_is_safe<const volatile Elem1*, const volatile Elem2*, Pr> == false);
+    // contiguous iterators to volatile should explode
+    STATIC_ASSERT(_Equal_memcmp_is_safe<typename vector<volatile Elem1>::iterator, typename vector<Elem2>::iterator,
```
Arg you are right, I did not want to use span as that was a hassle wrt to versions this is tested with as span is only C++20. Maybe only use array, which should be sufficient?

StephanTLavavej replied:
`array<volatile T, N>` is less weird (doesn't involve allocators) but I personally feel that it should still be forbidden. I think that given the pointer coverage in all modes, we'll be OK.

miscco commented on tests/std/tests/VSO_0180469_ptr_cat/test.cpp:
```diff
@@ -305,13 +391,16 @@ void equal_safe_test_cases() {
     test_case_Equal_memcmp_is_safe<false, base_class, base_class>();
 
+#ifdef __cpp_lib_byte
+    test_case_Equal_memcmp_is_safe<true, byte, byte>();
```
Nitpick: Should we add some failure test, e.g `test_case_Equal_memcmp_is_safe<false, char, byte>();`

StephanTLavavej replied:
Excellent idea! I'll push that.

StephanTLavavej commented:
I'm porting this to MSVC and will need to update that internal PR if any more changes are pushed here.

StephanTLavavej commented on stl/inc/xutility:
```diff
@@ -4657,55 +4657,93 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
-}
+template <class _Iter1, class _Iter2, class _Pr>
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe =
+    _Equal_memcmp_is_safe_helper<remove_const_t<_Iter1>, remove_const_t<_Iter2>, _Pr>;
```
😸=🐈+😊

StephanTLavavej commented:
Thanks for enhancing this optimization! Now, users of C++20 ranges will get the same codegen as for classic algorithms. 😸

miscco commented:
I would also like to thank me for pushing you to write awesome code

cpplearner commented:
This seems to cause DevCom-1263877 (`std::equal` does not work with custom contiguous iterators).

#1433 contains a fix, but a smaller PR might be preferable.

AdamBucior commented:
I don't think smaller PR would be preferable, #1433 is not huge.
